clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -660,6 +674,74 @@ public void subscribe(TopicPartition... partitions) {
     }
 
     /**
+     * Subscribes to topics matching specified pattern and uses the consumer's group
+     * management functionality. The pattern matching will be done periodically against topics
+     * existing at the time of check.
+     * <p>
+     * As part of group management, the consumer will keep track of the list of consumers that
+     * belong to a particular group and will trigger a rebalance operation if one of the
+     * following events trigger -
+     * <ul>
+     * <li>Number of partitions change for any of the subscribed list of topics
+     * <li>Topic is created or deleted
+     * <li>An existing member of the consumer group dies
+     * <li>A new member is added to an existing consumer group via the join API
+     * </ul>
+     *
+     * @param pattern Pattern to subscribe to
+     */
+    @Override
+    public void subscribe(Pattern pattern) {
+        acquire();
+        try {
+            log.debug("Subscribed to pattern: {}", pattern);
+            this.subscriptions.subscribe(pattern);
+
+            if (this.patternSubscriptionTask == null) {
+                schedulePatternSubscriptionTask(this.patternSubscriptionIntervalMs);
+            }
+        } finally {
+            release();
+        }
+    }
+
+    private void schedulePatternSubscriptionTask(final long interval) {
+        this.patternSubscriptionTask = new PatternSubscriptionTask(interval);
+        client.schedule(this.patternSubscriptionTask, time.milliseconds() + interval);
+    }
+
+    private boolean matchesSubscribedPattern(String topic) {
+        return this.subscriptions.getSubscribedPattern().matcher(topic).matches();
+    }
+
+    private boolean isAlreadySubscribed(String topic) {
+        return this.subscriptions.subscribedTopics().contains(topic);
+    }
+
+    /**
+     * Unsubscribe from topics subscribed via pattern subscription
+     *
+     * @param pattern Pattern to unsubscribe from, must match the pattern subscribed to
+     */
+    public void unsubscribe(Pattern pattern) {
```

Comment:
Would it have to? I would see each call to subscribe as replacing the previous one. Perhaps if we wanted to support multiple regex subscriptions, we could do:
```java
void subscribe(List<Pattern> patterns);
```
And we could require users to use '^' for blacklisting. This would be consistent with the changes proposed in KAFKA-2388.